openedx/core/djangoapps/password_policy/compliance.py

```diff
@@ -0,0 +1,140 @@
+    """
+    pass
+
+
+class NonCompliantPasswordWarning(Exception):
+    """
+    Exception that should be raised when a user who will soon be required to be compliant with password policy
+    requirements is found to have a non-compliant password.
+    """
+    pass
+
+
+def should_enforce_compliance_on_login():
+    """
+    Returns a boolean indicating whether or not password policy compliance should be enforced on login.
+    """
+    config = _rollout_config()
+    return config.get('ENFORCE_COMPLIANCE_ON_LOGIN', False)
+
+
+def enforce_compliance_on_login(user, password):
+    """
+    Verify that the user's password is compliant with password policy rules and determine what should be done
+    if it is not.
+
+    Raises NonCompliantPasswordException when the password is found to be non-compliant and the compliance deadline
+    for the user has been reached. In this case, login should be prevented.
+
+    Raises NonCompliantPasswordWarning when the password is found to be non-compliant and the compliance deadline for
+    the user is in the future.
+
+    Returns None when the password is found to be compliant, or when no deadline for compliance has been set for the
+    user.
+
+    Important: This method should only be called AFTER the user has been authenticated.
+    """
+    is_compliant = _check_user_compliance(user, password)
+    if is_compliant:
+        return
+
+    deadline = _get_compliance_deadline_for_user(user)
+    if deadline is None:
+        return
+
+    now = datetime.now(pytz.UTC)
+    if now >= deadline:
+        raise NonCompliantPasswordException(
+            _(
+                '{platform_name} now requires more complex passwords. Your current password does not meet the new '
+                'requirements. Change your password now to continue using the site. Thank you for helping us keep '
+                'your data safe.'
+            ).format(platform_name=settings.PLATFORM_NAME)
+        )
+    else:
+        raise NonCompliantPasswordWarning(
+            _(
+                '{platform_name} now requires more complex passwords. Your current password does not meet the new '
+                'requirements. You must change your password by {deadline} to be able to continue using the site. '
+                'Thank you for helping us keep your data safe.'
+            ).format(
+                platform_name=settings.PLATFORM_NAME,
+                deadline=strftime_localized(deadline, DEFAULT_SHORT_DATE_FORMAT)
+            )
+        )
+
+
+def _rollout_config():
+    """
+    Return a dictionary with configuration settings for managing the rollout of password policy compliance
+    enforcement.
+    """
+    return getattr(settings, 'PASSWORD_POLICY_COMPLIANCE_ROLLOUT_CONFIG', {})
+
+
+def _check_user_compliance(user, password):
+    """
+    Returns a boolean indicating whether or not the user is compliant with password policy rules.
+    """
+    try:
+        validate_password(password, user=user)
```

Comment:
PasswordHistory users will hit a problem with this, because of password-reuse (current password equal to old password).

Reply:
We now skip those checks.